### PR TITLE
Don't add the same method twice to an unsolved class

### LINKED ISSUE #185: Minimized File Fails to Compile
The following logs are provided whilst compiling Specimin's minimized files in ASHE. We are receiving a **100% failure rate** from what I have experience in bceul-util.

`--args=--outputDirectory "my/temp/dir" --root "my/root/bcel-util/src/main/java" --targetFile "org/plumelib/bcelutil/NoConstraintsVisitor.java" --targetMethod "org.plumelib.bcelutil.NoConstraintsVisitor#setFrame(Frame)"`

15:11:55.915 [main] INFO  edu.njit.jerse.ashe.services.SpeciminTool - 
Specimin output:

> Configure project :
No explicit dependency on the Checker Framework found, using default version 3.42.0

> Task :createCheckerFrameworkManifest UP-TO-DATE
> Task :spotlessInternalRegisterDependencies UP-TO-DATE
> Task :spotlessJava UP-TO-DATE
> Task :spotlessJavaApply UP-TO-DATE
> Task :spotlessMisc UP-TO-DATE
> Task :spotlessMiscApply UP-TO-DATE
> Task :spotlessApply UP-TO-DATE
> Task :compileJava UP-TO-DATE
> Task :processResources UP-TO-DATE
> Task :classes UP-TO-DATE
> Task :run

BUILD SUCCESSFUL in 1s
9 actionable tasks: 1 executed, 8 up-to-date

15:11:55.916 [main] INFO  edu.njit.jerse.ashe.utils.JavaCodeCorrector - Target file minimized successfully.
15:11:55.916 [main] INFO  edu.njit.jerse.ashe.services.SpeciminTool - Compiling minimized file: /my/temp/dir/org/plumelib/bcelutil/NoConstraintsVisitor.java
15:11:56.187 [main] INFO  edu.njit.jerse.ashe.services.SpeciminTool - 
Specimin output:
/my/temp/dir/org/plumelib/bcelutil/NoConstraintsVisitor.java:3: error: package org.apache.bcel.verifier.structurals does not exist
import org.apache.bcel.verifier.structurals.Frame;
                                           ^
/my/temp/dir/org/plumelib/bcelutil/NoConstraintsVisitor.java:4: error: package org.apache.bcel.verifier.structurals does not exist
import org.apache.bcel.verifier.structurals.InstConstraintVisitor;
                                           ^
/my/temp/dir/org/plumelib/bcelutil/NoConstraintsVisitor.java:6: error: cannot find symbol
public class NoConstraintsVisitor extends InstConstraintVisitor {
                                          ^
  symbol: class InstConstraintVisitor
/my/temp/dir/org/plumelib/bcelutil/NoConstraintsVisitor.java:9: error: cannot find symbol
    public void setFrame(Frame f) {
                         ^
  symbol:   class Frame
  location: class NoConstraintsVisitor
/my/temp/dir/org/plumelib/bcelutil/NoConstraintsVisitor.java:8: error: method does not override or implement a method from a supertype
    @Override
    ^
5 errors

15:11:56.187 [main] ERROR edu.njit.jerse.ashe.services.SpeciminTool - Minimized file failed to compile. Exit code: 1
15:11:56.188 [main] INFO  edu.njit.jerse.ashe.Ashe - Skipping...

### PR DESCRIPTION
**Pull request by kelloggm**
Fixes #185

### CODE REVIEW

**Review by LoiNguyenCS — CHANGES_REQUESTED** [review]
Professor,

Given that this PR is about recognizing different expressions that belongs to the same method, maybe it is more efficient to rewrite the `equal` method in `UnsolvedMethod` instead? That way we can reuse the codes for other contexts.

File: src/main/java/org/checkerframework/specimin/UnsolvedClassOrInterface.java [hunk 1]
```diff
@@ -187,6 +187,50 @@ public Set<String> getClassFields() {
...
+      // Skip methods that are definitely different.
+      if (!method.getName().equals(otherMethod.getName())
+          || otherMethod.getParameterList().size() != paramList.size()) {
+        continue;
+      }
+
+      for (int i = 0; i < paramList.size(); ++i) {
+        String paramType = paramList.get(i);
+        String otherParamType = otherMethod.getParameterList().get(i);
+        if ((this.packageName + "." + otherParamType).equals(paramType)) {
```
Comment (LoiNguyenCS) [inline src/main/java/org/checkerframework/specimin/UnsolvedClassOrInterface.java:206]: Is `this.packageName` the package of the current class? How are we sure that it is also the package of the name of the related parameter?

Comment (kelloggm) [reply src/main/java/org/checkerframework/specimin/UnsolvedClassOrInterface.java:206]: > Is `this.packageName` the package of the current class?

Yes. In particular, it's the name of the class that the method is being added to.

> How are we sure that it is also the package of the name of the related parameter?

We're not, I suppose. But we're relying on the fact that a simple name `Bar` used as a parameter in a class in package `com.foo` will be interpreted as `com.foo.Bar`.

**Comment by kelloggm** [comment]
> Given that this PR is about recognizing different expressions that belongs to the same method, maybe it is more efficient to rewrite the equal method in `UnsolvedMethod` instead? That way we can reuse the codes for other contexts.

I considered this idea, but decided against it because `UnsolvedMethod` doesn't know the fully-qualified name of the class that the method belongs to.

File: src/main/java/org/checkerframework/specimin/UnsolvedClassOrInterface.java (hunk 1 above)
Comment (LoiNguyenCS) [reply src/main/java/org/checkerframework/specimin/UnsolvedClassOrInterface.java:206]: Thanks for the explanation.

**Comment by LoiNguyenCS** [comment]
> I considered this idea, but decided against it because UnsolvedMethod doesn't know the fully-qualified name of the class that the method belongs to.

I see. Thank you.

**Review by LoiNguyenCS — APPROVED** [review]
LGTM

**Pull request merged**